### PR TITLE
Define delimiter-encapsulation

### LINKED ISSUE #65: Specification errata
I've posted a response to [a thread](https://discourse.citationstyles.org/t/should-if-else-blocks-use-the-delimiter-from-a-parent-group/1511/9?u=frank_bennett) by @cormacrelf on CSL Discourse in which he raises questions over some of the language in the CSL Specification. The issue should be thrashed out further on Discourse to be sure there is consensus, but I'll put down a marker here on portions of the spec that I think could justifiably be amended without a version increment.

The main item is the [classification of `cs:choose` as a "rendering element"](https://github.com/citation-style-language/documentation/blob/2f6161950734da35fe0538ed5b6f6da0be260136/specification.txt#L1392). The history is laid out in the comment linked above, but the short version is that this appears to be a change made in an early draft of CSL 1.0.1, the implications of which (for the treatment of inherited delimiters) were not clear at the time.

A second issue is [this sentence](https://github.com/citation-style-language/documentation/blob/2f6161950734da35fe0538ed5b6f6da0be260136/specification.txt#L1369) in the **Group** description:

> "``cs:group`` may carry the ``delimiter`` attribute to separate its *child elements*, as well as ``affixes`` and ``display`` attributes (applied to the output of the group as a whole) and ``formatting`` attributes *(transmitted to the enclosed elements)*." (emphases added)

Charles Parnot noted in 2012 that "child elements" here intends to refer to descendants, and Sylvester Kiel noted in 2014 that "transmitted to the enclosed elements" is not the intended behavior. As these interpretations were agreed at the time, an amendment along those lines would make it clear that current implementations have not arbitrarily diverged from the specification.

### PR DESCRIPTION
**Pull request by cormacrelf**
Fixes #65, Supersedes #116.

There are more words but I think it's easier to understand, especially with an example. It's also easier to extend it to other sequence-type elements as they arise.

### CODE REVIEW

**Comment by bdarcus** [comment]
Copying @jgm if he wants to weigh in on this alternative text.

See also @cormacrelf's [longer explanation](https://github.com/citation-style-language/documentation/pull/116#issuecomment-685708451) on the initial PR.

**Comment by bwiernik** [comment]
This is much clearer, I agree.

**Comment by denismaier** [comment]
I also agree that this is concise and clear.

**Review by bdarcus — CHANGES_REQUESTED** [review]

File: specification.rst [hunk 1]
```diff
@@ -2239,6 +2249,24 @@ allowed when ``cs:date`` calls a localized date format), ``cs:names``
 (delimiting names within name lists), ``cs:group`` and ``cs:layout`` (delimiting
 the output of the child elements).
 
+A delimiting element is any element as above which takes a ``delimiter`` 
+attribute, whether the attribute is supplied or not.
+
+The output of all delimiting elements is delimiter-encapsulated: no 
+``delimiter`` from a parent delimiting element will delimit between multiple 
+outputs of an inner delimiter-encapsulated element. The following produces
```
[the remark this answers is not in the document]
Comment (bwiernik) [reply specification.rst:2257]: We've avoided using "shall produce"-type language. When something is required, we either use "must" or just the present tense "produces"

**Comment by bdarcus** [comment]
Seems a little weird to me, but OK; as long as consistent.

>

File: specification.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdarcus) [reply specification.rst:2257]: Seems a little weird to me, but OK; as long as consistent.h

[the remark this answers is not in the document]
Comment (cormacrelf) [reply specification.rst:2257]: I said inner because it's not necessarily a direct child. `group > choose > if > group` is covered as well. So you could say "outer" instead of "parent".

**Comment by bdarcus** [comment]
Ancestor/descendent would be the more generic XML language.

**Comment by bdarcus** [comment]
You want to weigh in @PaulStanley or @andras-simonyi?

**Review by bdarcus — APPROVED** [review]

File: specification.rst (hunk 1 above)
Comment (bdarcus) [inline specification.rst:2257]:
```suggestion
``delimiter`` from an ancestor delimiting element will delimit between multiple 
outputs of a descendant delimiter-encapsulated element. The following produces
```

**Comment by bdarcus** [comment]
@adam3smith - if you have a moment, any feedback on this?

File: specification.rst [hunk 2]
```diff
@@ -2239,6 +2249,24 @@ allowed when ``cs:date`` calls a localized date format), ``cs:names``
...
+.. sourcecode:: xml
+
+    <group delimiter="; ">
+      <text value="a" />
+      <group>
+        <text value="b" />
+        <text value="c" />
+      </group>
+    </group>
+
```
Comment (adam3smith) [inline specification.rst:2269]:
```suggestion
``retrieved: <http://example.com``:

.. sourcecode:: xml

    <group delimiter=": ">
      <text term="retrieved" />
      <group>
        <text term="&lt;" />
        <text variable="URL"/>
      </group>
    </group>

```

Comment (adam3smith) [reply specification.rst:2269]: Happy for other examples, but the specifications use realistic examples throughout and I think we should stick with that both for consistency and for readability

**Review by adam3smith — COMMENTED** [review]
One suggestion on the example. 
I like the clarifications, so generally +1 from me. 

I don't love the term "delimiter-encapsulated" -- I don't think it's intelligible without the definition and I think ideally it should be at least somewhat self-explanatory (I also think it sounds a bit awkward). 

Do we absolutely need a term? Could we say something like "the output of a macro is not subject to surrounding delimiters" and something equivalent for delimiting elements? Not wedded to this change, but as a reminder, the principal audience of this document are style developers, so legibility is important.

**Comment by bwiernik** [comment]
> Do we absolutely need a term? Could we say something like "the output of a macro is not subject to surrounding delimiters" and something equivalent for delimiting elements? Not wedded to this change, but as a reminder, the principal audience of this document are style developers, so legibility is important.

How about
> the output of a macro does not inherit delimiters from ancestor elements

**Comment by bdarcus** [comment]
Shouldn't the primary audience for the spec be developers?

**Comment by bwiernik** [comment]
> Shouldn't the primary audience for the spec be developers?

Style writers read the spec much much more than developers do considering the comparative numbers of the two groups.

**Comment by bdarcus** [comment]
I don't really think that's relevant. I'm not saying it shouldn't serve
both audiences, but I strongly believe the primary audience should be
developers.

**Comment by adam3smith** [comment]
We've had this discussion a couple of months ago: There is a lot of small things that the current specs don't capture that a developer needs to do, currently captured in the test suite. If we add all of this to the main specs document, it becomes completely unusable for style authors, so one of the groups needs a separate document for it to be useful. My recollection was that we agreed on keeping the specs oriented towards authors and working towards separate developer specifications (maybe call them implementation notes or so).

File: specification.rst (hunk 2 above)
Comment (cormacrelf) [reply specification.rst:2269]: Agree.

**Comment by cormacrelf** [comment]
@adam3smith Agree on the terminology too. It made it shorter as well as a bit easier for authors. Thanks!

File: specification.rst [hunk 3]
```diff
@@ -384,6 +384,9 @@ the item type is "book":
       </citation>
     </style>
 
+The output of a ``<text macro="...">`` does not have delimiters from a parent
```
Comment (adam3smith) [inline specification.rst:387]: should this be `parent` or `ancestor` (honestly not sure)?

File: specification.rst [hunk 4]
```diff
@@ -1543,6 +1546,10 @@ control the testing logic, with allowed values:
 -  "none" - element only tests "true" when none of the conditions test "true"
    for any given test value
 
+The output of ``cs:choose`` (i.e. the output of the matching ``cs:if``, 
+``cs:else-if``, or ``cs:else``) *does* have delimiters from a parent delimiting
```
Comment (adam3smith) [inline specification.rst:1550]: ditto

Comment (cormacrelf) [reply specification.rst]: In each case ancestor would be technically correct if you did `group with delimiter > choose > if > choose > if` hmm

Comment (cormacrelf) [reply specification.rst]: Nearest ancestor delimiting element does it I think

Comment (cormacrelf) [reply specification.rst]: Because I defined delimiting element to be even if there is no delimiter set on it, yeah

**Comment by cormacrelf** [comment]
I think that's done

File: specification.rst [hunk 5]
```diff
@@ -384,6 +384,9 @@ the item type is "book":
       </citation>
     </style>
 
+The output of a ``<text macro="...">`` does not have delimiters from any
```
Comment (adam3smith) [inline specification.rst:387]: Thanks for working on this! I think we're almost there. I think we can improve on interspersed. 
Does "The output of a ``<text macro="...">`` does not take delimiters from any
ancestor delimiting element" (and similarly below) capture it? 
"Interspersed" sounds like a random process to my (admittedly non-native speaker) ears

Comment (bdarcus) [reply specification.rst:387]: I'm lost at this point. Can you guys "resolve" the suggestions or comments that you've addressed, once you have, so that we can tell when ready to merge?

Comment (adam3smith) [reply specification.rst:387]: From my end this is the only remaining unresolved issue. I've left your proposed change unresolved (since it's yours)

**Comment by bwiernik** [comment]
> We've had this discussion a couple of months ago: There is a lot of small things that the current specs don't capture that a developer needs to do, currently captured in the test suite. If we add all of this to the main specs document, it becomes completely unusable for style authors, so one of the groups needs a separate document for it to be useful. My recollection was that we agreed on keeping the specs oriented towards authors and working towards separate developer specifications (maybe call them implementation notes or so).

Yes, that was my understanding too.

File: specification.rst (hunk 5 above)
Comment (bwiernik) [inline specification.rst:387]:
```suggestion
Delimiters from any ancestor delimiting element are not applied within the output of <text macro="..."> element (see delimiter).
```

How about that?

Comment (adam3smith) [reply specification.rst:387]: I like it (though I think merging the suggestion won't work since it only replaces the first line.

**Comment by jgm** [comment]
> We've had this discussion a couple of months ago: There is a lot of small things that the current specs don't capture that a developer needs to do, currently captured in the test suite. If we add all of this to the main specs document, it becomes completely unusable for style authors, so one of the groups needs a separate document for it to be useful. My recollection was that we agreed on keeping the specs oriented towards authors and working towards separate developer specifications (maybe call them implementation notes or so).

I wasn't involved in this discussion, but my 2 cents as a developer:  trying to guess why the test suite requires things that aren't mentioned explicitly in the spec is no fun.  It makes sense to keep the spec manageable, but why not create a separate document with all the additional information that developers might need?

**Comment by bdarcus** [comment]
I think the primer should be the style author facing document, and the spec
should be developer facing.

File: specification.rst [hunk 6]
```diff
@@ -384,6 +384,9 @@ the item type is "book":
...
+ancestor delimiting element interspersed within it (see `delimiter`_).
```
Comment (bwiernik) [inline specification.rst:388]:
```suggestion
Delimiters from any ancestor delimiting element are not applied within the output of <text macro="..."> element (see delimiter).
```

File: specification.rst [hunk 7]
```diff
@@ -1543,6 +1545,10 @@ control the testing logic, with allowed values:
 -  "none" - element only tests "true" when none of the conditions test "true"
    for any given test value
 
+The output of ``cs:choose`` (i.e. the output of the matching ``cs:if``,
+``cs:else-if``, or ``cs:else``) *does* have delimiters from the nearest
+ancestor delimiting element interspersed within it (see `delimiter`_).
+
```
Comment (bwiernik) [inline specification.rst:1551]:
```suggestion
Delimiters from the nearest delimiters from the nearest ancestor delimiting element *are* applied within the output of ``cs:choose`` (i.e., the output of the matching ``cs:if``, ``cs:else-if``, or ``cs:else``; see `delimiter`_).

```

File: specification.rst [hunk 8]
```diff
@@ -2239,6 +2245,24 @@ allowed when ``cs:date`` calls a localized date format), ``cs:names``
 (delimiting names within name lists), ``cs:group`` and ``cs:layout`` (delimiting
 the output of the child elements).
 
+A delimiting element is any element as above which takes a ``delimiter`` 
+attribute, whether the attribute is supplied or not.
```
Comment (bwiernik) [inline specification.rst:2249]:
```suggestion
A delimiting element is any element as above which takes a ``delimiter`` attribute, whether the attribute is supplied or not.
```

File: specification.rst [hunk 9]
```diff
@@ -2239,6 +2245,24 @@ allowed when ``cs:date`` calls a localized date format), ``cs:names``
...
+
+The output of a delimiting element does not have delimiters from any ancestor
+delimiting element interspersed within it. The following produces ``retrieved:
+<http://example.com>``:
```
Comment (bwiernik) [inline specification.rst:2253]:
```suggestion
Delimiters from any ancestor delimiting element are not applied within the output of a delimiting element. The following produces ``retrieved: <http://example.com>``:
```

File: specification.rst (hunk 8 above)
Comment (bwiernik) [reply specification.rst:2249]: Putting single sentences on single lines.

File: specification.rst (hunk 7 above)
Comment (bwiernik) [reply specification.rst:1551]: Using the "delimiters are applied within" phrasing instead of "output has delimiters interspersed" phrasing.

File: specification.rst (hunk 9 above)
Comment (bwiernik) [reply specification.rst:2253]: Same "interspersed" comment as above.

**Review by bwiernik — APPROVED** [review]
This looks good to me. A minor formatting change and a suggestion for wording other than "interspersed".

**Review by adam3smith — APPROVED** [review]
Lgtm now, thanks!

**Comment by adam3smith** [comment]
> but my 2 cents as a developer: trying to guess why the test suite requires things that aren't mentioned explicitly in the spec is no fun. It makes sense to keep the spec manageable, but why not create a separate document with all the additional information that developers might need?

Absolutely; I don't think there's any disagreement that having citeproc developers rely purely on the test suite isn't great for all sorts of reasons, so a document that exhaustively describes expected citeproc behavior in precise language is definitely needed. I think the only reason it doesn't exist yet is the labor involved.

> I think the primer should be the style author facing document, and the spec should be developer facing.

I don't care about the names of the respective documents at all, so if there's a good reason to call the citeproc dev documents "specificiation," by all means let's do so. I don't think the technical document is ready, though, and I don't think there's any reason to delay the 1.0.2 release while waiting for it.

(The current primer is insufficient for style authoring. What style authors need is a document that exhaustively covers the syntax of CSL, including all variables and terms, and the general behavior of CSL, without too much implementation detail, e.g. things like handling double spaces & punctuation, a number of the other edge cases covered in the test suite currently.)
